packages/client/src/generation/GenericsArgsInfo.ts

```diff
@@ -0,0 +1,104 @@
+import { DMMF } from '../runtime/dmmf-types'
+import { Cache } from './Cache'
+import { getType } from './utils'
+
+type ToVisitItem = {
+  key: string
+  type: DMMF.SchemaArgInputType
+  parent?: ToVisitItem
+}
+
+export class GenericArgsInfo {
+  private _cache = new Cache<string, boolean>()
+
+  /**
+   * Determines if arg types need generic <$PrismaModel> argument added.
+   * Essentially, performs breadth-first search for any fieldRefTypes that
+   * do not have corresponding `meta.source` defined.
+   *
+   * @param type
+   * @returns
+   */
+  needsGenericModelArg(topLevelType: DMMF.SchemaArgInputType): boolean {
```

Comment:
Could it have simplified things to have a real recursion of a memoized function?

Reply:
Tried that before and not really: I think I severely underestimated amount of DMMF types, created by relations. With recursive version, benchmark from this PR runs out of stack space already with around 30 models.

Reply:
What I do think will make it easier to follow is separating iterative traversal code and actual logic for determining the need for generic. Then, with leaving out a couple of edge-case, this function could be simplified to something along the lines of:

```js
let currentAnswer = false
const typesStack =  []
traverse(startingType, {
   enterInputType(type) {
      currentAnswer = false
   },
   
   enterArg(arg) {
       if (arg.location === 'fieldRefTypes') {
           currentAnswer = true
       }
   }),
   
   leaveInputType(type) {
      this._cache.set(type, currentAnswer)
   }
})
```

However, I've decided not to spend the time on such helper for now, as pull request is already quite big, generic `traverse` helper will take more time and overall complexity of PR won't really decrease. I do think it will make sense to do it eventually, but I'd suggest we take time for that after patch release.

Reply:
Thanks for the explanation.